File: results/dnd-rest-benchmark/lifecycle-runs/20260731T051216Z_claude_claude-sonnet-5_php-stdlib/lib/play_downtime.php
<?php
declare(strict_types=1);

// Routes: Play (campaign-scoped recurring downtime activities that campaign
// members allocate to owned characters and progress repeatedly)
// ---------------------------------------------------------------------------

function validate_downtime_activity_payload($body): ?array {
    if ($body === null
        || !isset($body['activity_id'], $body['name'], $body['cycles_required'])
        || !is_string($body['activity_id']) || $body['activity_id'] === ''
        || !is_string($body['name']) || $body['name'] === ''
        || !is_valid_int_range($body['cycles_required'], 1, 10)) {
        return null;
    }

    return [
        'activity_id' => $body['activity_id'],
        'name' => $body['name'],
        'cycles_required' => (int)$body['cycles_required'],
    ];
}

if ($method === 'POST' && preg_match('#^/v1/play/campaigns/([^/]+)/downtime/activities$#', $path, $pm)) {
    $campaignId = $pm[1];
    $actor = require_actor($db);

    $campaign = require_play_campaign($db, $campaignId);

    if ($actor['username'] !== $campaign['owner']) {
        forbidden('only the campaign dm may create downtime activities');
    }

    $body = read_json_body();
    $validated = validate_downtime_activity_payload($body);
    if ($validated === null) {
        bad_request();
    }
    $activityId = $validated['activity_id'];

    $stmt = $db->prepare('SELECT 1 FROM play_campaign_downtime_activities WHERE campaign_id = ? AND activity_id = ?');
    $stmt->execute([$campaignId, $activityId]);
    if ($stmt->fetch(PDO::FETCH_ASSOC) !== false) {
        conflict('activity id already exists in this campaign');
    }

    $activity = [
        'activity_id' => $activityId,
        'name' => $validated['name'],
        'cycles_required' => $validated['cycles_required'],
    ];

    $stmt = $db->prepare('SELECT COALESCE(MAX(seq), 0) + 1 FROM play_campaign_downtime_activities WHERE campaign_id = ?');
    $stmt->execute([$campaignId]);
    $seq = (int)$stmt->fetchColumn();

    $stmt = $db->prepare('INSERT INTO play_campaign_downtime_activities (campaign_id, activity_id, seq, data) VALUES (?, ?, ?, ?)');
    $stmt->execute([$campaignId, $activityId, $seq, json_encode($activity)]);

    send_json($activity, 201);
}

if ($method === 'POST' && preg_match('#^/v1/play/campaigns/([^/]+)/characters/([^/]+)/downtime/allocations$#', $path, $pm)) {
    $campaignId = $pm[1];
    $charId = $pm[2];
    $actor = require_actor($db);

    require_play_campaign($db, $campaignId);

    $found = find_play_character($db, $campaignId, $charId);
    if ($found === null) {
        not_found('character not found');
    }
    [$username, $member] = $found;

    $owner = play_character_owner($db, $campaignId, $charId, $username);
    if ($actor['username'] !== $owner) {
        forbidden('only the character owner may allocate downtime for this character');
    }

    $body = read_json_body();
    if ($body === null || !isset($body['activity_id']) || !is_string($body['activity_id']) || $body['activity_id'] === '') {
        bad_request();
    }
    $activityId = $body['activity_id'];

    $stmt = $db->prepare('SELECT 1 FROM play_campaign_downtime_activities WHERE campaign_id = ? AND activity_id = ?');
    $stmt->execute([$campaignId, $activityId]);
    if ($stmt->fetch(PDO::FETCH_ASSOC) === false) {
        not_found('activity not found');
    }

    $stmt = $db->prepare('SELECT 1 FROM play_campaign_downtime_allocations WHERE campaign_id = ? AND character_id = ? AND activity_id = ?');
    $stmt->execute([$campaignId, $charId, $activityId]);
    if ($stmt->fetch(PDO::FETCH_ASSOC) !== false) {
        conflict('allocation already exists for this character and activity');
    }

    $allocation = [
        'character_id' => $charId,
        'activity_id' => $activityId,
        'cycles_completed' => 0,
        'completions' => 0,
    ];

    $stmt = $db->prepare('INSERT INTO play_campaign_downtime_allocations (campaign_id, character_id, activity_id, data) VALUES (?, ?, ?, ?)');
    $stmt->execute([$campaignId, $charId, $activityId, json_encode($allocation)]);

    send_json($allocation, 201);
}

if ($method === 'POST' && preg_match('#^/v1/play/campaigns/([^/]+)/characters/([^/]+)/downtime/allocations/([^/]+)/progress$#', $path, $pm)) {
    $campaignId = $pm[1];
    $charId = $pm[2];
    $activityId = $pm[3];
    $actor = require_actor($db);

    require_play_campaign($db, $campaignId);

    $found = find_play_character($db, $campaignId, $charId);
    if ($found === null) {
        not_found('character not found');
    }
    [$username, $member] = $found;

    $owner = play_character_owner($db, $campaignId, $charId, $username);
    if ($actor['username'] !== $owner) {
        forbidden('only the character owner may progress downtime for this character');
    }

    $stmt = $db->prepare('SELECT data FROM play_campaign_downtime_activities WHERE campaign_id = ? AND activity_id = ?');
    $stmt->execute([$campaignId, $activityId]);
    $activityRow = $stmt->fetch(PDO::FETCH_ASSOC);
    if ($activityRow === false) {
        not_found('activity not found');
    }
    $activity = json_decode($activityRow['data'], true);

    $stmt = $db->prepare('SELECT data FROM play_campaign_downtime_allocations WHERE campaign_id = ? AND character_id = ? AND activity_id = ?');
    $stmt->execute([$campaignId, $charId, $activityId]);
    $row = $stmt->fetch(PDO::FETCH_ASSOC);
    if ($row === false) {
        not_found('allocation not found');
    }
    $allocation = json_decode($row['data'], true);

    $allocation['cycles_completed'] = (int)$allocation['cycles_completed'] + 1;
    if ($allocation['cycles_completed'] >= (int)$activity['cycles_required']) {
        $allocation['cycles_completed'] = 0;
        $allocation['completions'] = (int)$allocation['completions'] + 1;
    }

    $stmt = $db->prepare('UPDATE play_campaign_downtime_allocations SET data = ? WHERE campaign_id = ? AND character_id = ? AND activity_id = ?');
    $stmt->execute([json_encode($allocation), $campaignId, $charId, $activityId]);

    send_json($allocation, 200);
}

if ($method === 'GET' && preg_match('#^/v1/play/campaigns/([^/]+)/characters/([^/]+)/downtime/allocations/([^/]+)$#', $path, $pm)) {
    $campaignId = $pm[1];
    $charId = $pm[2];
    $activityId = $pm[3];
    require_actor($db);

    require_play_campaign($db, $campaignId);

    $found = find_play_character($db, $campaignId, $charId);
    if ($found === null) {
        not_found('character not found');
    }

    $stmt = $db->prepare('SELECT 1 FROM play_campaign_downtime_activities WHERE campaign_id = ? AND activity_id = ?');
    $stmt->execute([$campaignId, $activityId]);
    if ($stmt->fetch(PDO::FETCH_ASSOC) === false) {
        not_found('activity not found');
    }

    $stmt = $db->prepare('SELECT data FROM play_campaign_downtime_allocations WHERE campaign_id = ? AND character_id = ? AND activity_id = ?');
    $stmt->execute([$campaignId, $charId, $activityId]);
    $row = $stmt->fetch(PDO::FETCH_ASSOC);
    if ($row === false) {
        not_found('allocation not found');
    }

    send_json(json_decode($row['data'], true), 200);
}

File: results/dnd-rest-benchmark/lifecycle-runs/20260730T075550Z_pi_kimi-k2p7-code_ruby-sinatra/lib/routes/play_downtime.rb
# frozen_string_literal: true

# Route definitions for recurring downtime activities and allocations.

# --- Downtime activities ---

post '/v1/play/campaigns/:id/downtime/activities' do
  content_type :json
  campaign_id = params[:id]
  validate_campaign_id!(campaign_id)

  username = require_dm_actor!
  campaign = load_play_campaign!(campaign_id)
  require_play_campaign_owner!(campaign, username)

  body = parse_json_body
  validate_downtime_activity_body!(body)

  json_error(409, 'activity already exists') if Storage.downtime_activity_exists?(campaign_id, body['activity_id'])

  activity = Storage.create_downtime_activity(campaign_id, body['activity_id'], body['name'], body['cycles_required'])

  status 201
  JSON.dump(activity)
end

# --- Downtime allocations ---

post '/v1/play/campaigns/:id/characters/:character_id/downtime/allocations' do
  content_type :json
  campaign_id = params[:id]
  character_id = params[:character_id]
  validate_campaign_id!(campaign_id)

  username = require_player_actor!

  campaign = load_play_campaign!(campaign_id)
  json_error(403, 'forbidden') unless Storage.play_campaign_member_exists?(campaign_id, username)

  member = Storage.load_play_campaign_member_by_character_id(campaign_id, character_id)
  json_error(404, 'character not found') unless member

  json_error(403, 'forbidden') unless member[:owner] == username

  body = parse_json_body
  validate_downtime_allocation_body!(body)

  activity_id = body['activity_id']
  json_error(404, 'activity not found') unless Storage.downtime_activity_exists?(campaign_id, activity_id)

  json_error(409, 'allocation already exists') if Storage.downtime_allocation_exists?(campaign_id, character_id, activity_id)

  allocation = Storage.create_downtime_allocation(campaign_id, character_id, activity_id)

  status 201
  JSON.dump(allocation)
end

post '/v1/play/campaigns/:id/characters/:character_id/downtime/allocations/:activity_id/progress' do
  content_type :json
  campaign_id = params[:id]
  character_id = params[:character_id]
  activity_id = params[:activity_id]
  validate_campaign_id!(campaign_id)

  username = require_player_actor!

  campaign = load_play_campaign!(campaign_id)
  json_error(403, 'forbidden') unless Storage.play_campaign_member_exists?(campaign_id, username)

  member = Storage.load_play_campaign_member_by_character_id(campaign_id, character_id)
  json_error(404, 'character not found') unless member

  json_error(403, 'forbidden') unless member[:owner] == username

  json_error(404, 'activity not found') unless Storage.downtime_activity_exists?(campaign_id, activity_id)
  json_error(404, 'allocation not found') unless Storage.downtime_allocation_exists?(campaign_id, character_id, activity_id)

  allocation = Storage.progress_downtime_allocation(campaign_id, character_id, activity_id)
  json_error(404, 'allocation not found') unless allocation

  status 200
  JSON.dump(allocation)
end

get '/v1/play/campaigns/:id/characters/:character_id/downtime/allocations/:activity_id' do
  content_type :json
  campaign_id = params[:id]
  character_id = params[:character_id]
  activity_id = params[:activity_id]
  validate_campaign_id!(campaign_id)

  user = authenticate_actor!
  username = user[:username]

  campaign = load_play_campaign!(campaign_id)
  require_play_campaign_access!(campaign, username)

  json_error(404, 'character not found') unless Storage.play_campaign_character_exists?(campaign_id, character_id)
  json_error(404, 'activity not found') unless Storage.downtime_activity_exists?(campaign_id, activity_id)

  allocation = Storage.load_downtime_allocation(campaign_id, character_id, activity_id)
  json_error(404, 'allocation not found') unless allocation

  JSON.dump(allocation)
end
